File: pec-observation-app/android-app/app/src/main/kotlin/br/gov/educacao/sp/pecobservation/adapters/ui/viewmodels/RecordingViewModel.kt
package br.gov.educacao.sp.pecobservation.adapters.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.gov.educacao.sp.pecobservation.application.usecases.RecordAudioUseCase
import br.gov.educacao.sp.pecobservation.ports.AudioRecordingPort
import br.gov.educacao.sp.pecobservation.ports.RecordingState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class RecordingViewModel @Inject constructor(
    private val recorder: AudioRecordingPort,
    private val useCase: RecordAudioUseCase,
) : ViewModel() {

    val state: StateFlow<RecordingState> = recorder.state
        .stateIn(viewModelScope, SharingStarted.Eagerly, RecordingState.IDLE)

    val elapsedSeconds: StateFlow<Int> = recorder.elapsedSeconds
        .stateIn(viewModelScope, SharingStarted.Eagerly, 0)

    var error: String? = null
        private set

    fun startRecording(observationId: UUID) {
        viewModelScope.launch {
            try {
                useCase.startRecording(observationId)
                error = null
            } catch (e: Exception) {
                error = e.message
            }
        }
    }

    fun pauseRecording() {
        viewModelScope.launch { recorder.pauseRecording() }
    }

    fun resumeRecording() {
        viewModelScope.launch { recorder.resumeRecording() }
    }

    fun stopRecording(observationId: UUID, onDone: () -> Unit = {}) {
        viewModelScope.launch {
            try {
                useCase.stopAndSave(observationId)
                error = null
                onDone()
            } catch (e: Exception) {
                error = e.message
            }
        }
    }

    fun cancelRecording() {
        recorder.cancelRecording()
    }

    fun formattedElapsed(seconds: Int): String {
        val m = seconds / 60
        val s = seconds % 60
        return "%02d:%02d".format(m, s)
    }
}

File: pec-observation-app/ios-app/Sources/PECObservation/Adapters/UI/ViewModels/RecordingViewModel.swift
#if os(iOS)
import Foundation
import Observation

@Observable
public final class RecordingViewModel {
    private let recordAudioUseCase: RecordAudioUseCase
    public var recordingState: RecordingState = .idle
    public var elapsedSeconds: Int = 0
    public var errorMessage: String?
    public var lastResult: AudioRecordingResult?

    public var formattedTime: String {
        let minutes = elapsedSeconds / 60
        let seconds = elapsedSeconds % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    public init(recordAudioUseCase: RecordAudioUseCase) {
        self.recordAudioUseCase = recordAudioUseCase
    }

    public func startRecording(observationId: UUID) async {
        errorMessage = nil
        do {
            try await recordAudioUseCase.startRecording(for: observationId)
            recordingState = .recording
        } catch {
            recordingState = .failed
            errorMessage = error.localizedDescription
        }
    }

    public func pauseRecording() async {
        do {
            try await recordAudioUseCase.pauseRecording()
            recordingState = .paused
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func resumeRecording() async {
        do {
            try await recordAudioUseCase.resumeRecording()
            recordingState = .recording
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func stopRecording(observationId: UUID) async {
        recordingState = .stopping
        do {
            lastResult = try await recordAudioUseCase.stopRecording(for: observationId)
            recordingState = .completed
        } catch {
            recordingState = .failed
            errorMessage = error.localizedDescription
        }
    }
}
#endif
